File: app/src/main/java/me/ykrank/s1next/view/fragment/PmGroupsFragment.java
package me.ykrank.s1next.view.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.PmGroup;
import me.ykrank.s1next.data.api.model.collection.PmGroups;
import me.ykrank.s1next.data.api.model.wrapper.BaseDataWrapper;
import me.ykrank.s1next.data.event.NoticeRefreshEvent;
import me.ykrank.s1next.util.L;
import me.ykrank.s1next.util.MathUtil;
import me.ykrank.s1next.view.adapter.BaseRecyclerViewAdapter;
import me.ykrank.s1next.view.adapter.PmGroupsRecyclerViewAdapter;
import me.ykrank.s1next.widget.EventBus;


public final class PmGroupsFragment extends BaseLoadMoreRecycleViewFragment<BaseDataWrapper<PmGroups>> {

    public static final String TAG = PmGroupsFragment.class.getName();
    private PmGroupsRecyclerViewAdapter mRecyclerAdapter;

    @Inject
    EventBus mEventBus;

    public static PmGroupsFragment newInstance() {
        PmGroupsFragment fragment = new PmGroupsFragment();
        return fragment;
    }

    @Override
    BaseRecyclerViewAdapter getRecyclerViewAdapter() {
        return mRecyclerAdapter;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        App.getAppComponent().inject(this);
        super.onViewCreated(view, savedInstanceState);
        L.leaveMsg("PmGroupsFragment");
        getActivity().setTitle(R.string.pms);

        RecyclerView recyclerView = getRecyclerView();
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mRecyclerAdapter = new PmGroupsRecyclerViewAdapter(getActivity());
        recyclerView.setAdapter(mRecyclerAdapter);
    }

    @Override
    boolean isCardViewContainer() {
        return true;
    }

    @Override
    Observable<BaseDataWrapper<PmGroups>> getSourceObservable(int pageNum) {
        return mS1Service.getPmGroups(pageNum);
    }

    @Override
    void onNext(BaseDataWrapper<PmGroups> data) {
        super.onNext(data);
        PmGroups pmGroups = data.getData();
        if (pmGroups.getPmGroupList() != null) {
            mRecyclerAdapter.diffNewDataSet(pmGroups.getPmGroupList(), false);
            // update total page
            setTotalPages(MathUtil.divide(pmGroups.getTotal(), pmGroups.getPmPerPage()));
        }

        if (getPageNum() == 1) {
            mEventBus.post(new NoticeRefreshEvent(data.getData().hasNew(), false));
        }
    }

    @NonNull
    @Override
    BaseDataWrapper<PmGroups> appendNewData(@Nullable BaseDataWrapper<PmGroups> oldData, @NonNull BaseDataWrapper<PmGroups> newData) {
        if (oldData != null) {
            List<PmGroup> oldPmGroups = oldData.getData().getPmGroupList();
            List<PmGroup> newPmGroups = newData.getData().getPmGroupList();
            if (newPmGroups == null) {
                newPmGroups = new ArrayList<>();
            }
            if (oldPmGroups != null) {
                newPmGroups.addAll(0, oldPmGroups);
            }
        }
        return newData;
    }
}

File: app/src/main/java/me/ykrank/s1next/view/fragment/PmGroupsFragment.kt
package me.ykrank.s1next.view.fragment

import android.os.Bundle
import android.view.View
import com.github.ykrank.androidtools.util.MathUtil
import com.github.ykrank.androidtools.widget.RxBus
import io.reactivex.Single
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.collection.PmGroups
import me.ykrank.s1next.data.api.model.wrapper.BaseDataWrapper
import me.ykrank.s1next.view.adapter.BaseRecyclerViewAdapter
import me.ykrank.s1next.view.adapter.PmGroupsRecyclerViewAdapter
import me.ykrank.s1next.view.event.NoticeRefreshEvent
import javax.inject.Inject


class PmGroupsFragment : BaseLoadMoreRecycleViewFragment<BaseDataWrapper<PmGroups>>() {
    private lateinit var mRecyclerAdapter: PmGroupsRecyclerViewAdapter

    @Inject
    internal lateinit var mRxBus: RxBus

    override val recyclerViewAdapter: BaseRecyclerViewAdapter
        get() = mRecyclerAdapter

    override val isCardViewContainer: Boolean
        get() = true

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        App.appComponent.inject(this)
        super.onViewCreated(view, savedInstanceState)
        leavePageMsg("PmGroupsFragment")
        activity?.setTitle(R.string.pms)

        val recyclerView = recyclerView
        recyclerView.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        mRecyclerAdapter = PmGroupsRecyclerViewAdapter(activity)
        recyclerView.adapter = mRecyclerAdapter
    }

    override fun getPageSourceObservable(pageNum: Int): Single<BaseDataWrapper<PmGroups>> {
        return mS1Service.getPmGroups(pageNum)
    }

    override fun onNext(data: BaseDataWrapper<PmGroups>) {
        super.onNext(data)
        val pmGroups = data.data
        val pmGroupList = pmGroups.list
        if (pmGroupList != null) {
            mRecyclerAdapter.diffNewDataSet(pmGroupList, false)
            // update total page
            setTotalPages(MathUtil.divide(pmGroups.count, pmGroups.perPage))
        }

        if (pageNum == 1) {
            mRxBus.post(NoticeRefreshEvent::class.java, NoticeRefreshEvent(data.data.hasNew(), null))
        }
    }

    override fun appendNewData(oldData: BaseDataWrapper<PmGroups>?, newData: BaseDataWrapper<PmGroups>): BaseDataWrapper<PmGroups> {
        if (oldData != null) {
            val oldPmGroups = oldData.data.list
            var newPmGroups = newData.data.list?.toMutableList()
            if (newPmGroups == null) {
                newPmGroups = ArrayList()
                newData.data.list = newPmGroups
            }
            if (oldPmGroups != null) {
                newPmGroups.addAll(0, oldPmGroups)
            }
        }
        return newData
    }

    companion object {

        val TAG = PmGroupsFragment::class.java.name

        fun newInstance(): PmGroupsFragment {
            return PmGroupsFragment()
        }
    }
}
